Topological Sorting (#93)
bobluppes/graaf: docs/docs/algorithms/topological-sort/topological-sort.md

---
sidebar_position: 1
---

# Topological sort algorithm
Topological sort algorithm processing DAG(directed acyclic graph) using DFS traversal.
Each vertex is visited only after all its dependencies are visited.
The runtime of the algorithm is `O(|V|+|E|)` and the memory consumption is `O(|V|)`.

[wikipedia](https://en.wikipedia.org/wiki/Topological_sorting)

## Syntax

```cpp

> template <typename V, typename E>
> [[nodiscard]] std::vector<edge_id_t> kruskal_minimum_spanning_tree(
>     const graph<V, E, graph_type::UNDIRECTED>& graph);

Reviewer: ```suggestion
template <typename V, typename E>
[[nodiscard]] std::vector<vertex_id_t> topological_sort(
    const graph<V, E, graph_type::DIRECTED>& graph);
```

Author: Oh, my bad, I missed this one.

```

> - **graph** The graph to traverse.

Reviewer: ```suggestion
- **graph** The directed graph to traverse.
```

Author: Alright

> - **return** Vector of vertices sorted in topological order. If the graph contains cycles, it returns an empty vector.

Reviewer: ```suggestion
- **return** Vector of vertices sorted in topological order. If the graph contains cycles, it returns std::nullopt.
```
